### ISSUE TITLE
3.1.x -> 3.2.x contains breaking change (`outerWidth` behavior)

### ISSUE THREAD

**Issue by jacobq** [jquery/jquery#3611]
Please pardon this lengthy issue report. If you think it should be split into multiple separate issues please let me know and I will attempt to do this, but I'm not sure how these observations relate.

This may be related to #3589 or the fixes for #3193 mentioned in [the 3.2.0 release notes](https://blog.jquery.com/2017/03/16/jquery-3-2-0-is-out/), but I'm not sure. (I don't think it's related to [the 3.0 breaking change](https://jquery.com/upgrade-guide/3.0/#breaking-change-outerwidth-or-outerheight-on-window-includes-scrollbar-width-height) since it worked in previous 3.x versions.)

## Problem 1: Firefox: `.width()` & `.outerWidth()` return `%` values instead of `px` values for inline elements

Tested on v45.8.0 (ESR) on Linux as well as 51.0.1 (32-bit) on Windows 10 (64-bit)

* [Working (v3.1.1)](http://output.jsbin.com/jaxuqob/6)
* [Broken (v3.2.0)](http://output.jsbin.com/jaxuqob/5)
* [Broken (latest)](http://output.jsbin.com/jaxuqob/11)

## Duplicate of #3571 ~Problem 2: Chrome: `.width()` returns `0` for inline elements~
* [Working (v3.1.1)](http://output.jsbin.com/jaxuqob/10)
* [Broken (v3.2.0)](http://output.jsbin.com/jaxuqob/9)
* [Working? (latest)](http://output.jsbin.com/jaxuqob/12)

## Problem 3: `.width()` / `.outerWidth()` gives different/wrong result on Chrome + Linux

This appears to be related to how padding is used in the calculation as well as to setting:

```
html {
  box-sizing: border-box;
}

*, *:before, *:after {
    box-sizing: inherit;
}
```

* [Working (v.3.1.1)](https://output.jsbin.com/jaxuqob/65)
* [Broken (v.3.2.0)](https://output.jsbin.com/jaxuqob/64)
* [Broken (latest)](https://output.jsbin.com/jaxuqob/62)

![jquery-working-311](https://cloud.githubusercontent.com/assets/1633459/24712508/af39edde-19e8-11e7-9f9e-e9518368cf59.png)
![jquery-broken-latest](https://cloud.githubusercontent.com/assets/1633459/24712509/af458950-19e8-11e7-8bc0-8d91005c4824.png)
![jquery-broken-firefox-latest](https://cloud.githubusercontent.com/assets/1633459/24712512/af563b56-19e8-11e7-82f8-6fc95337501f.png)
![jquery-windows-working-latest](https://cloud.githubusercontent.com/assets/1633459/24712511/af52f84c-19e8-11e7-8d29-d8c1f045680d.png)

## Background / Other Notes
This all started with [a problem I had](https://github.com/Dogfalo/materialize/issues/3532) using [Materialize](http://materializecss.com/). with the latest jQuery (3.2.1). The "tab indicator" suddenly started appearing too wide. You can see this [here](https://jsbin.com/jujeju/21/edit?html,js,output).
After digging into this a bit I found that:

* It appears to affect Chrome on Linux (x86_64), not Firefox (ESR / 48.5.0) on Linux or common Windows browsers.
* It works fine with jQuery 3.1.x but not 3.2.x (assuming semver compliance this should not happen)
* Using `.get(0).getBoundingClientRect().width` instead of `.outerWidth()` avoided the problem

**Comment by Krinkle** [comment]
Problem 2 was indeed a regression (width/height broken for inline elements). Reported as issue #3571, since then resolved in master and released as 3.2.1. (Although with notable exception that it is still broken on Android 4 - see issue #3602).

**Comment by jacobq** [comment]
@Krinkle Thanks for pointing that out. I updated my issue/comment to reflect this. I also spent some time this morning trying to figure out the root cause of the tabs issue (see notes at bottom of issue text) with the latest jQuery. I narrowed it down quite a bit and added links to jsbins as well as screenshots showing the discrepancies I observed in hopes that someone from the jQuery team will be able to identify what change caused this.

**Comment by timmywil** [comment]
@jacobq Thank you for the thorough analysis!

**Comment by TimePerformance** [comment]
Problem 4: outerHeight() returns wrong value for a <td> in IE 11 (Windows 10) but works fine in Chrome

**Comment by faller** [comment]
I have the same problem like `Problem 3`

**Comment by smasala** [comment]
@jacobq thanks for this (Problem 3). I can confirm it's the same problem under windows with FF 54.0.1 (64-bit) & Chrome 59.0.3071.115 (64-bit) and only for jQuery versions `>=3.2.0`

**Comment by smasala** [comment]
For an even weirder version of this problem in chrome.... it seems dependant on the url structure. 

Datatables plugin:
1. clone https://github.com/smasala/ColResize
2. npm install (make sure jQuery `3.2.0` or `3.2.1` is noted in package.json)
3. open in chrome -> `file:///C:/Temp/ColResize/examples/index.html`
4. make a copy of `index.html` -> `index1.html`
5. open in chrome ->`file:///C:/Temp/ColResize/examples/index1.html` - here it works. (but `index.html` won't)

Make this even weirder:
1. move both `index.html` and `index1.html` into a sub folder of `examples`.
2. now both `index.html` and `index1.html` will render correctly

**Comment by jacobq** [comment]
@smasala Just for fun I tried the exercise you shared but saw no differences between `index.html` and `index1.html`. Perhaps this is because of platform differences (I am running Linux 4.9 on x86_64, node v6.11.1) or maybe it was because I wasn't sure what to look for.

Nevertheless, I would strongly encourage you to simplify the portion of code that you're analyzing. When there are lots of parts at work it becomes very easy to become mislead (e.g. accidental caching, race conditions, dependence on inputs / user behavior, etc. can make one believe the problem is caused by something other than the root cause), and there are a number of nuances one has to deal with regarding how a browser will behave differently for `file://` URLs. Based on your description I suspect that the problem you're seeing is different from the ones described by the issue (and that it may not even be related to `jQuery`).

If using an older version of `jQuery` solves the problem in your application I would recommend doing that until a 3.3.x release is ready since it's most likely the best use of your time.

**Comment by smasala** [comment]
@jacobq You would have seen the error if it hadn't have worked :)

I'm not looking for a solution with my plugin, just thought I'd post my random findings :) However, I found this `issue` because I narrowed the problem down to one line... jquery's `outerWidth` and only when `box-sizing` is set, after updating jQuery to the latest version (>=3.2.0) but was unable to replicate the issue with jsfiddle or similar. So it's good to have your links and examples above - I'm not going crazy after all :)

I've just tried https://github.com/jquery/jquery/issues/3611#issuecomment-317771987 on another machine and under Windows 10 (instead of 7) + Chrome it works fine too. Very odd. Perhaps there is something wrong with my SSD on the other machine. Will have to try it on other Win 7 setup just out of curiosity and a sanity check :) 👍

**Comment by smasala** [comment]
@jacobq FYI :) works on other Win 7 / Chrome systems just fine. However, can be reproduced on my machine every time without fail. `3.1.1` works, `3.2.x` produces the exact problem as you described above (problem 3). Also the case in Chrome 60.0.3112.78. I give up :)

**Comment by gibson042** [comment]
a49578499ec88c9fbca59ff0c80b23b798a55d99 fixes everything except Problem 1.

**Comment by jacobq** [comment]
@gibson042 Thanks; this looks a lot better, though there is still a discrepancy. Any idea why my jsbin example (#3) gets all zeros with 3.1.1 but not with the latest (jQuery v3.2.2-pre 3cf1d14ccc63b8054ee977e90c72c5da07c80e59)? (In Chrome on Windows I see differences on the order of 0.48749542236328125)

**Comment by gibson042** [comment]
Yes, it's because 3.1.1 was using `getBoundingClientRect` but latest is actually calculating CSS boxes. Even though the latter can only approximate, the former is wrong because it is affected by transforms (scale, rotate, etc.).

**Comment by jacobq** [comment]
OK. That makes sense. I will plan on sticking with 3.1.1 for my
project then since it's not currently using any transforms.

**Comment by gibson042** [comment]
I'd recommend that you instead plan to upgrade and use `getBoundingClientRect` directly if that's what your use case needs.

**Comment by gibson042** [comment]
I suppose it would also be possible for us to use the fractional component from gBCR, but that would be a separate issue:
```js
val = elem[ "offset" + dimension[ 0 ].toUpperCase() + dimension.slice( 1 ) ];
if ( elem.getClientRects().length ) {
	val += ( elem.getBoundingClientRect()[ dimension[ 0 ] ] + 0.5 ) % 1 - 0.5;
}
```

**Comment by mgol** [comment]
@gibson042 That would be vulnerable to transforms as well, wouldn't it?

**Comment by gibson042** [comment]
Yes, but with a maximum error of half a pixel (i.e., no change from current) and recovering perfect accuracy in the absence of transforms.

**Comment by jacobq** [comment]
This looks like it is fixed in v3.3.x (i.e. in the jsbin we now have ow - gbcr < 1 again). Can anyone else confirm?
